In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
car = pd.read_csv('quikr_car.csv')

In [3]:
car.head()

name   company  year          Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007         80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006       4,25,000   
2              Maruti Suzuki Alto 800 Vxi    Maruti  2018  Ask For Price   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014       3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014       5,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
2  22,000 kms    Petrol  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel

In [4]:
car.shape

(892, 6)

In [5]:
car.describe()

name company  year          Price  kms_driven fuel_type
count          892     892   892            892         840       837
unique         525      48    61            274         258         3
top     Honda City  Maruti  2015  Ask For Price  45,000 kms    Petrol
freq            13     235   117             35          30       440

In [6]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        892 non-null    object
 1   company     892 non-null    object
 2   year        892 non-null    object
 3   Price       892 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 41.9+ KB


how to find out mistakes

In [7]:
car.year.unique()

array(['2007', '2006', '2018', '2014', '2015', '2012', '2013', '2016',
       '2010', '2017', '2008', '2011', '2019', '2009', '2005', '2000',
       '...', '150k', 'TOUR', '2003', 'r 15', '2004', 'Zest', '/-Rs',
       'sale', '1995', 'ara)', '2002', 'SELL', '2001', 'tion', 'odel',
       '2 bs', 'arry', 'Eon', 'o...', 'ture', 'emi', 'car', 'able', 'no.',
       'd...', 'SALE', 'digo', 'sell', 'd Ex', 'n...', 'e...', 'D...',
       ', Ac', 'go .', 'k...', 'o c4', 'zire', 'cent', 'Sumo', 'cab',
       't xe', 'EV2', 'r...', 'zest'], dtype=object)

we can observe that there are many garbage values are there instead of years and years value are in object form and we want it in integer form

In [8]:
car['Price'].unique()

array(['80,000', '4,25,000', 'Ask For Price', '3,25,000', '5,75,000',
       '1,75,000', '1,90,000', '8,30,000', '2,50,000', '1,82,000',
       '3,15,000', '4,15,000', '3,20,000', '10,00,000', '5,00,000',
       '3,50,000', '1,60,000', '3,10,000', '75,000', '1,00,000',
       '2,90,000', '95,000', '1,80,000', '3,85,000', '1,05,000',
       '6,50,000', '6,89,999', '4,48,000', '5,49,000', '5,01,000',
       '4,89,999', '2,80,000', '3,49,999', '2,84,999', '3,45,000',
       '4,99,999', '2,35,000', '2,49,999', '14,75,000', '3,95,000',
       '2,20,000', '1,70,000', '85,000', '2,00,000', '5,70,000',
       '1,10,000', '4,48,999', '18,91,111', '1,59,500', '3,44,999',
       '4,49,999', '8,65,000', '6,99,000', '3,75,000', '2,24,999',
       '12,00,000', '1,95,000', '3,51,000', '2,40,000', '90,000',
       '1,55,000', '6,00,000', '1,89,500', '2,10,000', '3,90,000',
       '1,35,000', '16,00,000', '7,01,000', '2,65,000', '5,25,000',
       '3,72,000', '6,35,000', '5,50,000', '4,85,000', '3,29,5

Here we in Price feature we have "Ask For Price" instead of price and datatype is in object form we want them in Integer without commas

In [9]:
car['kms_driven'].unique()

array(['45,000 kms', '40 kms', '22,000 kms', '28,000 kms', '36,000 kms',
       '59,000 kms', '41,000 kms', '25,000 kms', '24,530 kms',
       '60,000 kms', '30,000 kms', '32,000 kms', '48,660 kms',
       '4,000 kms', '16,934 kms', '43,000 kms', '35,550 kms',
       '39,522 kms', '39,000 kms', '55,000 kms', '72,000 kms',
       '15,975 kms', '70,000 kms', '23,452 kms', '35,522 kms',
       '48,508 kms', '15,487 kms', '82,000 kms', '20,000 kms',
       '68,000 kms', '38,000 kms', '27,000 kms', '33,000 kms',
       '46,000 kms', '16,000 kms', '47,000 kms', '35,000 kms',
       '30,874 kms', '15,000 kms', '29,685 kms', '1,30,000 kms',
       '19,000 kms', nan, '54,000 kms', '13,000 kms', '38,200 kms',
       '50,000 kms', '13,500 kms', '3,600 kms', '45,863 kms',
       '60,500 kms', '12,500 kms', '18,000 kms', '13,349 kms',
       '29,000 kms', '44,000 kms', '42,000 kms', '14,000 kms',
       '49,000 kms', '36,200 kms', '51,000 kms', '1,04,000 kms',
       '33,333 kms', '33,600 kms', '5,

Here we have kms unit with no which we have to seprate and again datatype is in object form instead of Integer

In [10]:
car['fuel_type'].unique()

array(['Petrol', 'Diesel', nan, 'LPG'], dtype=object)

In Fuel_type feature we have some nan values

## List of problems in dataset
- names are pretty inconsistent
- names have company names attached to it
- some names are spam like 'Maruti Ertiga showroom condition with' and 'Well mentained Tata Sumo'
- company: many of the names are not of any company like 'Used', 'URJENT', and so on.
- year has many non-year values
- year is in object. Change to integer
- Price has Ask for Price
- Price has commas in its prices and is in object
- kms_driven has object values with kms at last.
- It has nan values and two rows have 'Petrol' in them
- fuel_type has nan values

### cleaning Dataset

In [11]:
backup = car.copy()

In [12]:
car = car[car['year'].str.isnumeric()]

car.head()

name   company  year          Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007         80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006       4,25,000   
2              Maruti Suzuki Alto 800 Vxi    Maruti  2018  Ask For Price   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014       3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014       5,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
2  22,000 kms    Petrol  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel

In [13]:
car['year'] = car['year'].astype(int)

In [14]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 842 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        842 non-null    object
 1   company     842 non-null    object
 2   year        842 non-null    int64 
 3   Price       842 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: int64(1), object(5)
memory usage: 46.0+ KB


In [15]:
car = car[car['Price'] != "Ask For Price"]

In [16]:
car['Price'] = car['Price'].str.replace(',','').astype(int)

In [17]:
car['kms_driven'] = car['kms_driven'].str.split(" ").str.get(0).str.replace(',', '')

In [18]:
car = car[car['kms_driven'].str.isnumeric()]

car['kms_driven'] = car['kms_driven'].astype(int)

car

name   company  year   Price  \
0      Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000   
1                   Mahindra Jeep CL550 MDI  Mahindra  2006  425000   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  325000   
4          Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000   
6                                 Ford Figo      Ford  2012  175000   
..                                      ...       ...   ...     ...   
883              Maruti Suzuki Ritz VXI ABS    Maruti  2011  270000   
885               Tata Indica V2 DLE BS III      Tata  2009  110000   
886                    Toyota Corolla Altis    Toyota  2009  300000   
888                     Tata Zest XM Diesel      Tata  2018  260000   
889                      Mahindra Quanto C8  Mahindra  2013  390000   

     kms_driven fuel_type  
0         45000    Petrol  
1            40    Diesel  
3         28000    Petrol  
4         36000    Diesel  
6         41000    Diesel  
..          ...       ...  
883       50000    Petrol  
885       30000    Diesel  
886      132000    Petrol  
888       27000    Diesel  
889       40000    Diesel  

[817 rows x 6 columns]

In [19]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 817 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        817 non-null    object
 1   company     817 non-null    object
 2   year        817 non-null    int64 
 3   Price       817 non-null    int64 
 4   kms_driven  817 non-null    int64 
 5   fuel_type   816 non-null    object
dtypes: int64(3), object(3)
memory usage: 44.7+ KB


In [20]:
car[car['fuel_type'].isna()]

name company  year   Price  kms_driven fuel_type
132  Toyota Corolla  Toyota  2009  275000       26000       NaN

In [21]:
# Drop NaN
car[~car['fuel_type'].isna()]

name   company  year   Price  \
0      Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000   
1                   Mahindra Jeep CL550 MDI  Mahindra  2006  425000   
3    Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  325000   
4          Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000   
6                                 Ford Figo      Ford  2012  175000   
..                                      ...       ...   ...     ...   
883              Maruti Suzuki Ritz VXI ABS    Maruti  2011  270000   
885               Tata Indica V2 DLE BS III      Tata  2009  110000   
886                    Toyota Corolla Altis    Toyota  2009  300000   
888                     Tata Zest XM Diesel      Tata  2018  260000   
889                      Mahindra Quanto C8  Mahindra  2013  390000   

     kms_driven fuel_type  
0         45000    Petrol  
1            40    Diesel  
3         28000    Petrol  
4         36000    Diesel  
6         41000    Diesel  
..          ...       ...  
883       50000    Petrol  
885       30000    Diesel  
886      132000    Petrol  
888       27000    Diesel  
889       40000    Diesel  

[816 rows x 6 columns]

In [22]:
car['name'] = car['name'].str.split(' ').str.slice(0,3).str.join(' ')
car

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
3         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
4    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
6                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
883      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
885          Tata Indica V2      Tata  2009  110000       30000    Diesel
886    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
888            Tata Zest XM      Tata  2018  260000       27000    Diesel
889      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[817 rows x 6 columns]

In [23]:
car = car.reset_index(drop=True)

In [24]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817 entries, 0 to 816
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        817 non-null    object
 1   company     817 non-null    object
 2   year        817 non-null    int64 
 3   Price       817 non-null    int64 
 4   kms_driven  817 non-null    int64 
 5   fuel_type   816 non-null    object
dtypes: int64(3), object(3)
memory usage: 38.4+ KB


In [25]:
car.describe()

year         Price     kms_driven
count   817.000000  8.170000e+02     817.000000
mean   2012.440636  4.115503e+05   46250.714810
std       4.002354  4.749173e+05   34283.745254
min    1995.000000  3.000000e+04       0.000000
25%    2010.000000  1.750000e+05   27000.000000
50%    2013.000000  2.999990e+05   41000.000000
75%    2015.000000  4.900000e+05   56758.000000
max    2019.000000  8.500003e+06  400000.000000

we can see that min_price of car is 3lac, 25% car lies under 1.75 lac, 50% under 3lac and 75% under 4.9 lac but still max_price is around 85lac which clearly signifies that it is an outlier

In [26]:
car[car['Price'] > 6e6]

name   company  year    Price  kms_driven fuel_type
535  Mahindra XUV500 W6  Mahindra  2014  8500003       45000    Diesel

In [27]:
car = car[car['Price'] < 6e6].reset_index(drop=True)

In [28]:
car.to_csv('cleaned car data.csv')

## Model

In [29]:
x = car.drop(columns='Price')
y = car['Price']

In [30]:
x

name   company  year  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006          40    Diesel
2         Hyundai Grand i10   Hyundai  2014       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014       36000    Diesel
4                 Ford Figo      Ford  2012       41000    Diesel
..                      ...       ...   ...         ...       ...
811      Maruti Suzuki Ritz    Maruti  2011       50000    Petrol
812          Tata Indica V2      Tata  2009       30000    Diesel
813    Toyota Corolla Altis    Toyota  2009      132000    Petrol
814            Tata Zest XM      Tata  2018       27000    Diesel
815      Mahindra Quanto C8  Mahindra  2013       40000    Diesel

[816 rows x 5 columns]

In [31]:
y

0       80000
1      425000
2      325000
3      575000
4      175000
        ...  
811    270000
812    110000
813    300000
814    260000
815    390000
Name: Price, Length: 816, dtype: int64

In [32]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [39]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline

In [34]:
ohe = OneHotEncoder()

In [35]:
ohe.fit(x[['name', 'company', 'fuel_type']])

OneHotEncoder()

In [45]:
ohe.categories_

[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
        'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
        'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
        'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat Diesel',
        'Chevrolet Beat LS', 'Chevrolet Beat LT', 'Chevrolet Beat PS',
        'Chevrolet Cruze LTZ', 'Chevrolet Enjoy', 'Chevrolet Enjoy 1.4',
        'Chevrolet Sail 1.2', 'Chevrolet Sail UVA', 'Chevrolet Spark',
        'Chevrolet Spark 1.0', 'Chevrolet Spark LS', 'Chevrolet Spark LT',
        'Chevrolet Tavera LS', 'Chevrolet Tavera Neo', 'Datsun GO T',
        'Datsun Go Plus', 'Datsun Redi GO', 'Fiat Linea Emotion',
        'Fiat Petra ELX', 'Fiat Punto Emotion', 'Force Motors Force',
        'Force Motors One', 'Ford EcoSport', 'Ford EcoSport Ambiente',
        'Ford EcoSport Titanium', 'Ford EcoSport Trend',
        'Ford Endeavor 4x4', 'Ford Fiesta', 'Ford Fiesta SXi', 'Ford Figo',
        '

In [46]:
column_trans = make_column_transformer((OneHotEncoder(categories= ohe.categories_),['name', 'company', 'fuel_type']), remainder='passthrough')

In [41]:
linearReg = LinearRegression()

In [48]:
pipe = make_pipeline( column_trans,linearReg)

In [49]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
       'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
       'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
       'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat...
                                                                            array(['Audi', 'BMW', 'Chevrolet', 'Datsun', 'Fiat', 'Force', 'Ford',
       'Hindustan', 'Honda', 'Hyundai', 'Jaguar', 'Jeep', 'Land',
       'Mahindra', 'Maruti', 'Mercedes', 'Mini', 'Mitsubishi', 'Nissan',
       'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen', 'Volvo'],
      dtype=object),
                                                                            array(['Diesel', 'LPG', 'Pe

In [51]:
y_pred = pipe.predict(X_test)

In [52]:
y_pred

array([ 514857.52603909,  792188.8298291 ,  212782.11116645,
        436442.16898432,  373712.0011413 ,  278960.08870154,
        658873.36520638, 2223429.14336527,  477080.34097532,
        387681.28831062,  167004.83084677,   76540.86821327,
        431555.96625024,  359583.81529941,  168144.68477487,
        630151.53281837,  371119.57105893,  488841.92747495,
        508885.31990594,  326325.10099168,  138537.47116306,
        437206.94214106,  367799.79130888,  930861.57975258,
        436442.16898432,  305151.82653599,  574042.2110842 ,
        151055.6481813 ,  310922.23514123,  650675.93002575,
         61843.37177029,  556635.18182582,  558449.70944039,
        584529.34326336,  573075.80214292,  207345.80173367,
        465439.70230345,  344403.60017426,  370034.08943073,
        341583.0063042 ,  396919.10747499,  435475.76004303,
        255395.27710555,  191288.53953555,  216144.05536623,
        712127.49591606, -176182.14101294,  442889.452602  ,
        172779.3869044 ,

In [53]:
r2_score(y_test,y_pred)

0.7329147016947059